From ef3c88b68de63426e095d9f87fc8b7a8556425a9 Mon Sep 17 00:00:00 2001
From: Jacek Caban <jacek@codeweavers.com>
Date: Thu, 14 May 2026 22:57:59 +0200
Subject: [PATCH 07/48] ntdll: Add support for ARM64EC cooperative suspend.

When running simulated code, instead of suspending a thread immediately, the kernel
sets a registered doorbell, expecting the jit to stop execution as soon as it reaches
a consistent state that can be represented as an x86_64 context.

Based on patch by Billy Laws.

CW-Bug-Id: #27308

(cherry picked from commit b8d8f34ffd8b62be712c8839ed082b40f57f07c8)
---
 dlls/ntdll/unix/signal_arm64.c | 40 ++++++++++++++++++++++++++++++++--
 include/wine/server_protocol.h |  5 +++--
 server/protocol.def            |  5 +++--
 server/request_trace.h         |  1 +
 server/thread.c                | 14 ++++++++++--
 5 files changed, 57 insertions(+), 8 deletions(-)

diff --git a/dlls/ntdll/unix/signal_arm64.c b/dlls/ntdll/unix/signal_arm64.c
index 90d5d931e10..18426bc72bf 100644
--- a/dlls/ntdll/unix/signal_arm64.c
+++ b/dlls/ntdll/unix/signal_arm64.c
@@ -65,6 +65,18 @@ WINE_DEFAULT_DEBUG_CHANNEL(seh);
 #define NTDLL_DWARF_H_NO_UNWINDER
 #include "dwarf.h"
 
+struct arm64_thread_data
+{
+    BOOL suspend_pending;
+};
+
+C_ASSERT( sizeof(struct arm64_thread_data) <= sizeof(((struct ntdll_thread_data *)0)->cpu_data) );
+
+static inline struct arm64_thread_data *arm64_thread_data(void)
+{
+    return (struct arm64_thread_data *)ntdll_get_thread_data()->cpu_data;
+}
+
 /***********************************************************************
  * signal context platform-specific definitions
  */
@@ -332,7 +344,20 @@ static void restore_context( const CONTEXT *context, ucontext_t *sigcontext )
 NTSTATUS signal_set_full_context( CONTEXT *context )
 {
     struct syscall_frame *frame = get_syscall_frame();
-    NTSTATUS status = NtSetContextThread( GetCurrentThread(), context );
+    struct arm64_thread_data *arm64_data = arm64_thread_data();
+    NTSTATUS status;
+
+    if (arm64_data->suspend_pending)
+    {
+        sigset_t old_set;
+        pthread_sigmask( SIG_BLOCK, &server_block_set, &old_set );
+        *NtCurrentTeb()->ChpeV2CpuAreaInfo->SuspendDoorbell = 0;
+        arm64_data->suspend_pending = FALSE;
+        wait_suspend( context );
+        status = NtSetContextThread( GetCurrentThread(), context );
+        pthread_sigmask( SIG_SETMASK, &old_set, NULL );
+    }
+    else status = NtSetContextThread( GetCurrentThread(), context );
 
     if (!status && (context->ContextFlags & CONTEXT_INTEGER) == CONTEXT_INTEGER)
         frame->restore_flags |= CONTEXT_INTEGER;
@@ -1311,9 +1336,20 @@ static void quit_handler( int signal, siginfo_t *siginfo, void *sigcontext )
 static void usr1_handler( int signal, siginfo_t *siginfo, void *sigcontext )
 {
     ucontext_t *ucontext = sigcontext;
+    CHPE_V2_CPU_AREA_INFO *chpe;
     CONTEXT context;
 
-    if (is_inside_syscall( SP_sig(ucontext) ))
+    if ((chpe = NtCurrentTeb()->ChpeV2CpuAreaInfo) && chpe->InSimulation && chpe->SuspendDoorbell)
+    {
+        NTSTATUS status = server_select( NULL, 0, SELECT_INTERRUPTIBLE | SELECT_COOPERATIVE_SUSPEND,
+                                         0, NULL, NULL );
+        if (status == STATUS_THREAD_WAS_SUSPENDED)
+        {
+            *chpe->SuspendDoorbell = -1;
+            arm64_thread_data()->suspend_pending = TRUE;
+        }
+    }
+    else if (is_inside_syscall( SP_sig(ucontext) ))
     {
         context.ContextFlags = CONTEXT_FULL | CONTEXT_EXCEPTION_REQUEST;
         NtGetContextThread( GetCurrentThread(), &context );
diff --git a/include/wine/server_protocol.h b/include/wine/server_protocol.h
index dc1ceeb142e..d9087179099 100644
--- a/include/wine/server_protocol.h
+++ b/include/wine/server_protocol.h
@@ -1614,8 +1614,9 @@ struct select_reply
     /* VARARG(call,apc_call); */
     /* VARARG(contexts,contexts); */
 };
-#define SELECT_ALERTABLE     1
-#define SELECT_INTERRUPTIBLE 2
+#define SELECT_ALERTABLE           1
+#define SELECT_INTERRUPTIBLE       2
+#define SELECT_COOPERATIVE_SUSPEND 4
 
 
 
diff --git a/server/protocol.def b/server/protocol.def
index 03565fe4d53..ac38076b22c 100644
--- a/server/protocol.def
+++ b/server/protocol.def
@@ -1463,8 +1463,9 @@ struct cpu_topology_override
     VARARG(call,apc_call);     /* APC call arguments */
     VARARG(contexts,contexts); /* suspend context(s) */
 @END
-#define SELECT_ALERTABLE     1
-#define SELECT_INTERRUPTIBLE 2
+#define SELECT_ALERTABLE           1
+#define SELECT_INTERRUPTIBLE       2
+#define SELECT_COOPERATIVE_SUSPEND 4
 
 
 /* Create an event */
diff --git a/server/request_trace.h b/server/request_trace.h
index ec36688c3b3..5e94ddf90ef 100644
--- a/server/request_trace.h
+++ b/server/request_trace.h
@@ -4577,6 +4577,7 @@ static const struct
     { "SHUTDOWN_IN_PROGRESS",        STATUS_SHUTDOWN_IN_PROGRESS },
     { "SUSPEND_COUNT_EXCEEDED",      STATUS_SUSPEND_COUNT_EXCEEDED },
     { "THREAD_IS_TERMINATING",       STATUS_THREAD_IS_TERMINATING },
+    { "THREAD_WAS_SUSPENDED",        STATUS_THREAD_WAS_SUSPENDED },
     { "TIMEOUT",                     STATUS_TIMEOUT },
     { "TOO_MANY_OPENED_FILES",       STATUS_TOO_MANY_OPENED_FILES },
     { "UNSUCCESSFUL",                STATUS_UNSUCCESSFUL },
diff --git a/server/thread.c b/server/thread.c
index 7caf39b1d50..5c4a6bd0207 100644
--- a/server/thread.c
+++ b/server/thread.c
@@ -134,6 +134,7 @@ struct context
     struct object           obj;        /* object header */
     struct object          *sync;       /* sync object for wait/signal */
     unsigned int            status;     /* status of the context */
+    int                     cooperative;/* waiting for the cooperative suspend */
     struct context_data     regs[2];    /* context data */
 };
 #define CTX_NATIVE  0  /* context for native machine */
@@ -446,6 +447,7 @@ static inline void init_thread_structure( struct thread *thread )
 
 static inline int is_thread_suspended( struct thread *thread )
 {
+    if (thread->context && thread->context->cooperative) return 0;
     if (thread->suspend) return 1;
     return !thread->bypass_proc_suspend && thread->process->suspend;
 }
@@ -486,8 +488,9 @@ static struct context *create_thread_context( struct thread *thread )
 {
     struct context *context;
     if (!(context = alloc_object( &context_ops ))) return NULL;
-    context->sync   = NULL;
-    context->status = STATUS_PENDING;
+    context->sync        = NULL;
+    context->status      = STATUS_PENDING;
+    context->cooperative = 0;
     memset( &context->regs, 0, sizeof(context->regs) );
     context->regs[CTX_NATIVE].machine = native_machine;
 
@@ -1193,6 +1196,12 @@ static int check_wait( struct thread *thread )
     if ((wait->flags & SELECT_INTERRUPTIBLE) && !list_empty( &thread->system_apc ))
         return STATUS_KERNEL_APC;
 
+    if ((wait->flags & SELECT_COOPERATIVE_SUSPEND) && thread->context)
+    {
+        thread->context->cooperative = 1;
+        return STATUS_THREAD_WAS_SUSPENDED;
+    }
+
     /* Suspended threads may not acquire locks, but they can run system APCs */
     if (is_thread_suspended( thread )) return -1;
 
@@ -2058,6 +2067,7 @@ DECL_HANDLER(select)
             copy_context( &ctx->regs[CTX_WOW], wow_context, wow_context->flags & ~ctx->regs[CTX_WOW].flags );
         }
         ctx->status = STATUS_SUCCESS;
+        ctx->cooperative = 0;
         current->suspend_cookie = req->cookie;
         signal_sync( ctx->sync );
     }
-- 
2.55.0

